src/Network/Transport/TCP/Internal.hs

```diff
@@ -278,15 +295,30 @@ tryShutdownSocketBoth sock = void . tryIO $
 --
 -- Throws an I/O exception if the socket closes before the specified
 -- number of bytes could be read
-recvExact :: N.Socket        -- ^ Socket to read from
+recvExact :: TCPInternalSettings -- ^ Settings for specific TCP behaviour
+          -> N.Socket        -- ^ Socket to read from
           -> Word32          -- ^ Number of bytes to read
           -> IO [ByteString] -- ^ Data read
-recvExact sock len = go [] len
+recvExact settings sock len = go [] len
   where
     go :: [ByteString] -> Word32 -> IO [ByteString]
     go acc 0 = return (reverse acc)
     go acc l = do
       bs <- NBS.recv sock (fromIntegral l `min` smallChunkSize)
+      -- If forcing TCP_QUICKACK is enabled, set it after every recv()
+      -- (see also http://stackoverflow.com/questions/1615447/disable-tcp-delayed-acks).
+      -- Currently, this location is the only call to recv() in the package.
+      -- If other direct calls to `NBS.recv` are added anywhere, this
+      -- logic needs to be applied there as well!
+      case tcpInternalForceQuickAck settings of
+        Nothing -> return ()
+#ifdef linux_HOST_OS
+        Just forceVal -> do
+          let val = if forceVal then 1 else 0
+          N.setSocketOption sock (N.CustomSockOpt (6, 12)) val -- 6 is IPPROTO_TCP, 12 is TCP_QUICKACK
+#else
+        Just _forceVal -> return () -- Non-Linux OSs don't have TCP_QUICKACK.
+#endif
```

Comment:
If we can check if socket option is supported, we wouldn't need this conditional compilation.